src/hooks/routing/useRouterTrade.ts

```diff
@@ -0,0 +1,164 @@
+import { skipToken } from '@reduxjs/toolkit/query/react'
+import { Currency, CurrencyAmount, TradeType } from '@uniswap/sdk-core'
+// Importing just the type, so smart-order-router is lazy-loaded
+// eslint-disable-next-line no-restricted-imports
+import type { ChainId } from '@uniswap/smart-order-router'
+import { useRouterArguments } from 'hooks/routing/useRouterArguments'
+import useActiveWeb3React from 'hooks/useActiveWeb3React'
+import useDebounce from 'hooks/useDebounce'
+import useIsValidBlock from 'hooks/useIsValidBlock'
+import useIsWindowVisible from 'hooks/useIsWindowVisible'
+// TODO: double-check that we're removing these analytics/metrics packages from widget?
+// import { IMetric, MetricLoggerUnit, setGlobalMetric } from '@uniswap/smart-order-router'
+// import { sendTiming } from 'components/analytics'
+import { useStablecoinAmountFromFiatValue } from 'hooks/useUSDCPrice'
+import ms from 'ms.macro'
+import { useMemo } from 'react'
+import { useGetQuoteQuery } from 'state/routing/slice'
+import { GetQuoteResult, InterfaceTrade, TradeState } from 'state/routing/types'
+import { computeRoutes, transformRoutesToTrade } from 'state/routing/utils'
+
+import { AUTO_ROUTER_SUPPORTED_CHAINS, useAutoRouterSupported } from './clientSideSmartOrderRouter'
+
+export const INVALID_TRADE = { state: TradeState.INVALID, trade: undefined }
+
+/**
+ * Returns the best trade by invoking the routing api or the smart order router on the client
+ * @param tradeType whether the swap is an exact in/out
+ * @param amountSpecified the exact amount to swap in/out
+ * @param otherCurrency the desired output/payment currency
+ */
+export function useRouterTrade<TTradeType extends TradeType>(
+  tradeType: TTradeType,
+  routerUrl?: string,
+  amountSpecified?: CurrencyAmount<Currency>,
+  otherCurrency?: Currency
+): {
+  state: TradeState
+  trade: InterfaceTrade<Currency, Currency, TTradeType> | undefined
+} {
+  const autoRouterSupported = useAutoRouterSupported()
+  const isWindowVisible = useIsWindowVisible()
+  // Debounce is used to prevent excessive requests to SOR, as it is data intensive.
+  // Fast user actions (ie updating the input) should be debounced, but currency changes should not.
+  const [debouncedAmount, debouncedOtherCurrency] = useDebounce(
+    useMemo(() => [amountSpecified, otherCurrency], [amountSpecified, otherCurrency]),
+    200
+  )
+  const debouncedAmountSpecified = autoRouterSupported && isWindowVisible ? debouncedAmount : undefined
+
+  const [currencyIn, currencyOut]: [Currency | undefined, Currency | undefined] = useMemo(
+    () =>
+      tradeType === TradeType.EXACT_INPUT
+        ? [debouncedAmountSpecified?.currency, debouncedOtherCurrency]
+        : [debouncedOtherCurrency, debouncedAmountSpecified?.currency],
+    [debouncedAmountSpecified, debouncedOtherCurrency, tradeType]
+  )
+
+  const chainId = currencyIn?.chainId as ChainId
+  if (chainId && !AUTO_ROUTER_SUPPORTED_CHAINS.includes(chainId)) {
```

Comment:
should we make `useAutoRouterSupported()` take in a chainId and move this logic in there?

Reply:
going to do this in a separate future PR - re: https://github.com/Uniswap/widgets/pull/54#discussion_r917935671